docs/reference/programming_model.rst

```diff
@@ -0,0 +1,259 @@
+        rectangles filled with downward pointing arrows inside.
+
+  Hierarchy of thread groups.
+
+* The innermost grouping is called a warp, or a wavefront in ISA terms. A warp
+  is the most tightly coupled groups of threads, both physically and logically.
+
+  When referring to threads inside a warp, they may be called lanes, and the
+  integral value identifying them the lane ID. Lane IDs aren't quieried like
+  other thread IDs, but are user-calculated. As a consequence they are only as
+  multi-dimensional as the user interprets the calculated values to be.
+
+  The size of a warp is architecture dependent and always fixed. Warps are
+  signified by the set of communication primitives at their disposal, detailed
+  under :ref:`warp_cross_lane_functions`.
+
+* The middle grouping is called a block or thread block. The defining feature
+  of a block is that all threads in a block will share an instance of memory
+  which they may use to share data or synchronize with oneanother.
+
+  The size of a block is user-configurable but is maxmized by the queryable
+  capabilites of the executing hardware. The unique ID of the thread within a
+  block is 3-dimensional as provided by the API. When linearizing thread IDs
+  within a block, assume the "fast index" being dimension ``x``, followed by
+  the ``y`` and ``z`` dimensions.
+
+* The outermost grouping is called a grid. A grid manifests as a single
+  dispatch of kernels for execution. The unique ID of each block within a grid
+  is 3-dimensional, as provided by the API and is queryable by every thread
+  within the block.
+
+Cooperative Groups Thread Model
+-------------------------------------------------------------------------------
+
+The Cooperative Groups API introduces new APIs to launch, group, subdivide,
+synchronize and identify threads, as well as some predefined group-collective
+algorithms, but most importantly a matching threading model to think in terms
+of. It relaxes some of the restrictions of the :ref:`inherent_thread_model`
+imposed by the strict 1:1 mapping of architectural details to the programming
+model.
+
+The rich set of APIs introduced by Cooperative Groups allow the programmer
+to define their own groups based on run-time predicates, but a set of implicit
+groups manifest based on kernel launch parameters.
+
+The thread hiearchy abstraction of Cooperative Groups manifest as depicted in
+:numref:`coop_thread_hierarchy`.
+
+.. _coop_thread_hierarchy:
+
+.. figure:: ../data/reference/programming_model/thread_hierarchy_coop.svg
+  :alt: Diagram depicting nested rectangles of varying color. The outermost one
+        titled "Grid", inside sets of different sized rectangles layered on
+        oneanother titled "Block". Each "Block" containing sets of uniform
+        rectangles layered on oneanother titled "Warp". Each of the "Warp"
+        titled rectangles filled with downward pointing arrows inside.
```

Comment:
This alt text is not fully accurate, the outermost rectangle is called "Multi Grid", and there are no warps in this picture (which is good that way imo).
I think this is just a copy-paste mistake, but just to make sure: this should convey, that the grids can have different amounts and layouts of blocks (here the layout is actually important, compared to the first figure)

The remarks from the first one also mostly apply - the passive voice makes this harder to understand.

Reply:
Fixed